rest/data/commit_queue.go

```diff
@@ -280,6 +283,95 @@ func (pc *DBCommitQueueConnector) IsAuthorizedToPatchAndMerge(ctx context.Contex
 	return hasPermission, nil
 }
 
+// EnqueuePRToCommitQueue enqueues an item to the commit queue to test and merge a PR.
+func EnqueuePRToCommitQueue(ctx context.Context, env evergreen.Environment, sc Connector, info commitqueue.EnqueuePRInfo, isGithubHook bool) (*restModel.APIPatch, error) {
+	settings := env.Settings()
+	userRepo := UserRepoInfo{
+		Username: info.Username,
+		Owner:    info.Owner,
+		Repo:     info.Repo,
+	}
+	authorized, err := sc.IsAuthorizedToPatchAndMerge(ctx, settings, userRepo)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting user info from GitHub API")
+	}
+	if !authorized {
+		return nil, errors.Errorf("user '%s' is not authorized to merge", userRepo.Username)
+	}
+
+	pr, err := sc.GetGitHubPR(ctx, userRepo.Owner, userRepo.Repo, info.PR)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting PR from GitHub API")
+	}
+
+	if pr == nil || pr.Base == nil || pr.Base.Ref == nil {
+		return nil, errors.New("PR contains no base branch label")
+	}
+
+	cqInfo := restModel.ParseGitHubComment(info.CommitMessage)
+	baseBranch := *pr.Base.Ref
+	projectRef, err := model.FindOneProjectRefWithCommitQueueByOwnerRepoAndBranch(userRepo.Owner, userRepo.Repo, baseBranch)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting project for '%s:%s' tracking branch '%s'", userRepo.Owner, userRepo.Repo, baseBranch)
+	}
+	if projectRef == nil {
+		return nil, errors.Errorf("no project with commit queue enabled for '%s:%s' tracking branch '%s'", userRepo.Owner, userRepo.Repo, baseBranch)
+	}
+
+	if utility.FromBoolPtr(projectRef.CommitQueue.RequireSigned) {
+		err = checkSignedCommit(ctx, settings, userRepo, info.PR)
+		if err != nil {
+			sendErr := thirdparty.SendCommitQueueGithubStatus(evergreen.GetEnvironment(), pr, message.GithubStateFailure, "can't enqueue with unsigned commits", "")
```

Comment:
Alright, I read this GitHub status logic over again, and it seems like it would be nice to send the GitHub failure back to the user for all error return cases after we've retrieved and validated the PR info so we can send a response but before the item has been successfully put in the commit queue. We could reduce the amount of duplicate `thirdparty.SendCommitQueueGithubStatus` calls by moving everything L321-358 into a helper function and if that helper errors, call `SendCommitQueueGithubStatus`. Something like:

```go
apiPatch, err := tryEnqueueItemForPR(...)
if err != nil {
    if sendErr := thirdparty.SendCommitQueueGithubStatus(...) { log the send error }
    return nil, errors.Wrap(err, ...)
}

func tryEnqueueItemForPR(...) {
    // check signed, check required approval count, create patch, enqueue in commit queue
}
```

Reply:
Sounds good, `SendCommitQueueGithubStatus` has an error description input param, so I added a third return value to `tryEnqueueItemForPR` that corresponds with the type of failure encountered, since providing the full `err.Error()` could be a long string.